### PR TITLE
Changes to support better separation between applying global and type serialization

### PR DESCRIPTION
**Pull request by wneild**
Changes include:

- The addition of a global serializer which can be configured with a custom UserSerializer for adding many custom Kryo serializers
- Redesigned typed serializer to only take a single optional custom Kryo serializer
- Method for registering customized global serialization

### CODE REVIEW

**Comment by jerrinot** [comment]
@wneild: many thanks for your contribution. There are some integration tests failing, you can run them locally by calling `mvn clean test -Pit`

Some more thoughts: I think the separation between Global / Type-specific serializers is useful when you are creating a customer subclass. However I quite like to have a single out-of-the-box implementation - it makes the default user experience configuration smoother - just a single class to deal with. Smooth user experience is one of the core features of this project. 

Here is a proposal:
1. keep a single out-of-the-box `Serializer` - it can be registered as a global serializer or for a specific type. Also make this class final so I cannot be sub-classesed
2. Have 2 abstract classes: `AbstractGlobalUserSerializier` and `AbstractTypeSpecificUserSerializler` (or something like this)

What do you think?

**Comment by wneild** [comment]
@jerrinot thanks for the feedback! Good suggestion, I've fixed the tests I missed and made those changes. 

The only thing I'm not sure on is the constructor argument for `AbstractGlobalUserSerializer`, I don't think the `UserSerializer` is an appropriate argument. I was thinking the 'UserSerializationBuilder' might be a good fit for now, any thoughts?

File: subzero-core/src/main/java/info/jerrinot/subzero/AbstractSerializer.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.StreamSerializer;
+import info.jerrinot.subzero.internal.strategy.KryoStrategy;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+public abstract class AbstractSerializer<T> implements StreamSerializer<T>, HazelcastInstanceAware {
```
Comment (jerrinot) [inline subzero-core/src/main/java/info/jerrinot/subzero/AbstractSerializer.java]: this probably should be package protected

File: subzero-core/src/main/java/info/jerrinot/subzero/SubZero.java [hunk 2]
```diff
@@ -23,22 +23,38 @@ private SubZero() {
...
+     * Use SubZero as a global serializer.
+     *
+     * This method configures Hazelcast to delegate a class serialization to SubZero when the class
+     * has no explicit strategy configured.
+     *
+     * @param config Hazelcast configuration to inject SubZero into
+     * @param serializerClazz Class of global serializer implementation to use
+     * @return Hazelcast configuration.
+     */
+    public static <T> Config useAsGlobalSerializer(Config config, Class<? extends AbstractGlobalUserSerializer> serializerClazz) {
```
Comment (jerrinot) [inline subzero-core/src/main/java/info/jerrinot/subzero/SubZero.java]: I would not add this method. Methods in the `SubZero` class are for user-convenience. For a quick-start if the default SubZero is good enough for you. If you are implementing own serializer then you do not need the convenience.

I expect majority of users won't/shouldn't need to create custom subclasses. The extra methods here could be confusing for new users.

**Comment by jerrinot** [comment]
@wneild: thanks for the updated PR. It looks good! I made some minor comments, please let me know what do you think about them. 

Also +1 to `UserSerializationBuilder` Please go ahead and rename it.

File: subzero-core/src/main/java/info/jerrinot/subzero/SubZero.java (hunk 2 above)
Comment (wneild) [inline subzero-core/src/main/java/info/jerrinot/subzero/SubZero.java]: The way I see it, SubZero is all about providing convenience, so I'm not convinced about not having some convenience method to add custom serialization configuration. The alternative is forcing the user to write out what's already available internally to SubZero class when registering the custom global serializer to hazelcast, which seems a waste.

I couldn't comment on how widely used custom serialization is (probably about 10% of Kryo users going by github stars https://github.com/magro/kryo-serializers) but I will remove the methods if you feel strongly.

**Comment by jerrinot** [comment]
good work, many thanks!

**Pull request merged**